eta/core/serial.py

```diff
@@ -670,32 +705,325 @@ def from_dict(cls, d):
 
             # Parse reflectively
             cls = etau.get_class(d["_CLS"])
-            ele_cls = etau.get_class(d[cls._ELE_CLS_FIELD])
+
         else:
             # Validates the cls settings
             cls()
-            # Parse using provided class
-            ele_cls = cls._ELE_CLS
 
-        return cls(**{
-            cls._ELE_ATTR: [ele_cls.from_dict(dd) for dd in d[cls._ELE_ATTR]]
-        })
+
+class DirectoryContainer(Container):
+    '''Container that represents a list of serializable objects in a
+    directory on disk. Any accessing of an element in the list causes a READ
+    from disk and updating any elements in the list REQUIRES setting the item
+    explicitly (and causes a WRITE to disk).
+
+    This class cannot be instantiated directly.
+
+    A list of uuids is managed by the DirectoryContainer in _ELE_ATTR to keep
+    order and find elements in the directory. This list a part of
+    serialization. The data directory location is also serialized in
+    _ELE_DIR_ATTR.
+
+    USE WITH CARE.
+
+    Usage:
+        element = cls._ELE_CLS()
+        c = DirectoryContainerSubclass.from_json("...")
+
+        # All getters cause an IMMEDIATE READ from disk for each element
+        element = c[0]
+        in_memory_list = [e for e in c] # NOT recommended
+
+        # The setters cause an IMMEDIATE WRITE to disk
+        c[0] = element
+        c.append(element)
+        c.add(element)
+    '''
+
+    #
+    # The name of the attribute that will store the path to the json data files
+    #
+    # Subclasses MUST set this field
+    #
+    _ELE_DIR_ATTR = None
+
+    def __init__(self, **kwargs):
```

Comment:
I really like how the "JSON serializable part of this is just a lightweight list of uuids. Nice idea!

For `Big<Class>` objects, I think we should support `from_zip()` and `to_zip()` methods as well, so that one can easily get ahold of all the data.

To implement `to_zip("/path/to/output.zip")`, I would propose that you arrange the files in a tempdir with structure:
```py
dir/
    index.json
    <elements>/
        <uuid>.json
```
then zip it and delete the tempdir.

Reply:
How about having that directory structure for the actual backing dir?  (Then the functionality would be similar to a `LabeledDataset`.)

Reply:
I thought about recommending that as well. Would be curious to hear Ben's thoughts. It is perhaps convenient to have independent control over the backing directory. Of course, that would be true for `LabeledDataset` as well...

Reply:
As a `Serializable` object, my feeling is that having the "index" writable to any location using `write_json` is more consistent.